### PR TITLE
Update Elasticsearch indices when a modification is published

### PR DESCRIPTION
**Pull request by sampaccoud**
## Purpose

Elasticsearch indices are created and populated via a management command. 

Up to now, any modification to the pages that was done after running this command were not reflected in the search indices.

## Proposal

- [x] add a method on the course run to easily get its parent course,
- [x] refactor indexers to allow updating documents one-by-one in Elasticsearch,
- [x] add signals to trigger update of the impacted documents in Elasticsearch,
- [x] test the added signals.

### CODE REVIEW

File: src/richie/apps/courses/models/course.py [hunk 1]
```diff
@@ -399,6 +399,18 @@ def state(self):
             self.start, self.end, self.enrollment_start, self.enrollment_end
         )
 
+    def get_course(self):
+        """Get the course for this course run."""
+        node = self.extended_object.node
+        ancestors_paths = [
```
Comment (mbenadda) [inline src/richie/apps/courses/models/course.py]: This assignment might deserve a comment unless it's regular old DjangoCMS stuff

File: src/richie/apps/search/indexers/categories.py [hunk 2]
```diff
@@ -57,76 +57,78 @@ class CategoriesIndexer:
     ]
 
     @classmethod
-    def get_es_documents(cls, index, action):
+    def get_es_document_for_category(cls, category, index=None, action="index"):
         """
         Load all the categories from the Category model and format them for the
```
Comment (mbenadda) [inline src/richie/apps/search/indexers/categories.py]:
```suggestion
        Build an Elasticsearch document from the category instance.
```

File: src/richie/apps/search/indexers/organizations.py [hunk 3]
```diff
@@ -47,68 +47,71 @@ class OrganizationsIndexer:
...
+            extended_object__title_set__published=True,
+        ).distinct():
+            yield cls.get_es_document_for_organization(
+                organization, index=index, action=action
+            )
+
+    @classmethod
+    def get_es_document_for_organization(cls, organization, index=None, action="index"):
         """
         Load all the organizations from the Organization model and format them for the
```
Comment (mbenadda) [inline src/richie/apps/search/indexers/organizations.py]: Same comment as in categories

File: src/richie/apps/search/indexers/organizations.py [hunk 4]
```diff
@@ -47,68 +47,71 @@ class OrganizationsIndexer:
     display_fields = ["absolute_url", "logo", "title.*"]
 
     @classmethod
-    def get_es_documents(cls, index, action):
+    def get_es_documents(cls, index=None, action="index"):
```
Comment (mbenadda) [inline src/richie/apps/search/indexers/organizations.py]: This is a nitpick but in courses and categories you put `get_es_document_for_course` first

File: src/richie/apps/search/signals.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+"""Update Elasticsearch indexes each time a page is modified."""
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import transaction
+
+from richie.apps.courses.models import Category, Course, Organization
+from richie.apps.search.indexers import ES_INDICES
+from richie.apps.search.index_manager import richie_bulk
+
+
+def update_course(course, _language):
```
Comment (mbenadda) [inline src/richie/apps/search/signals.py]: Nice design for all these `update_` methods 👍

File: src/richie/apps/courses/models/course.py (hunk 1 above)
Comment (sampaccoud) [reply src/richie/apps/courses/models/course.py]: It's inspired from treebeard's `get_ancestors` method: https://github.com/django-treebeard/django-treebeard/blob/master/treebeard/mp_tree.py#L1052.

Maybe I can directly use it... let me try.

File: src/richie/apps/search/indexers/organizations.py (hunk 4 above)
Comment (sampaccoud) [reply src/richie/apps/search/indexers/organizations.py]: nice catch! I wanted to do it and forgot.

File: src/richie/apps/search/signals.py (hunk 5 above)
Comment (sampaccoud) [reply src/richie/apps/search/signals.py]: pylint did like it :disappointed: Let me try another design...

Comment (sampaccoud) [reply src/richie/apps/search/signals.py]: Done.

File: src/richie/apps/courses/models/course.py (hunk 1 above)
Comment (sampaccoud) [reply src/richie/apps/courses/models/course.py]: Done. Much clearer.

File: src/richie/apps/search/signals.py (hunk 5 above)
Comment (mbenadda) [reply src/richie/apps/search/signals.py]: I think the previous one was more elegant but this is alright too

**Pull request merged**